ibis/expr/deferred.py

```diff
@@ -0,0 +1,121 @@
+from __future__ import annotations
+
+import functools
+import operator
+from typing import TYPE_CHECKING, Any, Callable, TypeVar
+
+import toolz
+
+if TYPE_CHECKING:
+    import ibis.expr.types as ir
+
+
+class Deferred:
+    """Construct a deferred expression."""
+
+    __slots__ = ("resolve",)
+
+    def __init__(self, resolve: Callable = toolz.identity) -> None:
+        self.resolve = resolve
+
+    def _wrap(self, func: Callable, *args: Any, **kwargs: Any) -> Deferred:
+        """Wrap `func` in a `Deferred` instance."""
+        return self.__class__(
+            lambda expr, func=func, self=self, args=args, kwargs=kwargs: func(
```

Comment:
Wouldn't it be better to define a local function for better readability?

Reply:
Something like: 

```python
def resolve(expr):
    expr = self.result(expr)
    args = (_resolve(arg, expr=expr) for arg in args)
    kwargs = {name: _resolve(arg, expr=expr) for name, arg in kwargs.items()}
    return func(expr, *args, **kwargs)
```

Reply:
Roger, done.